api/src/classes/shop.py

```diff
@@ -0,0 +1,294 @@
+
+            # Sort by given type
+            if params.sort_by is not None:
+                for option in params.sort_by:
+                    match option:
+                        case "price":
+                            q = q.order_by(desc(ShopItem.price)) if not params.ascending else q.order_by(ShopItem.price)
+
+                        case "quantity":
+                            q = (
+                                q.order_by(desc(ShopItem.available_quantity))
+                                if not params.ascending
+                                else q.order_by(ShopItem.available_quantity)
+                            )
+
+                        case _:
+                            continue
+
+            paginator: Paginate = Paginate(query=q, session=session, params=params)
+
+            res: list[ShopItemPublic] = [ShopItemPublic.model_validate(x) for x in paginator.get_data()]
+
+            if not res:
+                raise HTTPException(status_code=404, detail="Shop Items not found.")
+
+            paginator.add_data({"shop_items": res})
+            return paginator.get_page()
+
+        def purchase_item(self, company: Company, data: ShopItemPurchase) -> None:  # noqa: PLR0912
+            """Add purchased Item to the Company's inventory.
+
+            :param data: Data on Shop Item purchase.
+            :param company: Target company purchasing the Shop Item.
+            :return: None
+            """
+            # Check if the item is not disabled.
+            if self.shop_item.is_disabled:
+                raise HTTPException(status_code=400, detail="Cannot purchase item; disabled.")
+
+            # Check if the Company is not bankrupt
+            if company.is_bankrupt:
+                raise HTTPException(status_code=403, detail="Company is bankrupt; cannot purchase.")
+
+            # Check that the Item has stock
+            if self.shop_item.available_quantity < 1:
+                raise HTTPException(status_code=403, detail="No available items to purchase.")
+
+            # Check that the Company is not buying more than what is available.
+            if self.shop_item.available_quantity < data.purchase_quantity:
+                raise HTTPException(status_code=400, detail="Cannot purchase more than what is available.")
+
+            # Check that the User can purchase the Item.
+            purchase_cost: float = self.shop_item.price * data.purchase_quantity
+            if purchase_cost > company.networth:
+                raise HTTPException(status_code=400, detail="Cannot purchase; insufficient funds.")
+
+            # Purchase the item(s)
+            company_inventory: list[Inventory] | None = company.inventory
+
+            if len(company_inventory) == 0:
+                # Add Item to Company Inventory
+                if company.id:
+                    new_inventory_item: Inventory = Inventory(
```

Comment:
The `Inventory` typehint isn't needed but if you'd prefer we can keep it. IDE's should be able to automatically infer its type from the assignment